appserver/jdbc/jdbc-config/src/main/java/org/glassfish/jdbc/config/validators/JdbcConnectionPoolValidator.java

```diff
@@ -71,17 +72,58 @@ public boolean isValid(final ResourcePool pool,
 
         if(poolFaults == ConnectionPoolErrorMessages.MAX_STEADY_INVALID) {
             if(pool instanceof JdbcConnectionPool) {
+                
+                Properties systemProps = System.getProperties();
+                
                 JdbcConnectionPool jdbcPool = (JdbcConnectionPool) pool;
                 String maxPoolSize = jdbcPool.getMaxPoolSize();
                 String steadyPoolSize = jdbcPool.getSteadyPoolSize();
-                if(steadyPoolSize == null) {
+                
+                int maxPoolSizeValue = 0;
+                int steadyPoolSizeValue = 0;
+                
+                if (steadyPoolSize == null) {
                     steadyPoolSize = Constants.DEFAULT_STEADY_POOL_SIZE;
+                } else if(steadyPoolSize.startsWith("$")) {
+                    //Try to retrive system variable value
+                    steadyPoolSize = steadyPoolSize.substring(2, steadyPoolSize.length() - 1);
+                    
+                    if (systemProps.getProperty(steadyPoolSize) == null || systemProps.getProperty(steadyPoolSize).isEmpty()) {
+                        //System property doesn't exist so try environment variables
+                        if (System.getenv(steadyPoolSize) == null || System.getenv(steadyPoolSize).isEmpty()) {
+                            //Environment variable doesn't exist so return false and log reasoning as error extension
+                            return false;
+                        }
+                        steadyPoolSize = System.getenv(steadyPoolSize);
+                    } else {
+                        steadyPoolSize = systemProps.getProperty(steadyPoolSize);
+                    }
                 }
+                
                 if (maxPoolSize == null) {
                     maxPoolSize = Constants.DEFAULT_MAX_POOL_SIZE;
+                } else if(maxPoolSize.startsWith("$")) {
```

Comment:
These blocks seem pretty similar, maybe it's possible to extract it into a method to make it a bit more generic? Or perhaps only partly, some of the environment and system variable checks could be